src/fheroes2/battle/battle_interface.cpp

```diff
@@ -1260,29 +1265,57 @@ void Battle::TurnOrder::redraw( const Unit * current, const uint8_t currentUnitC
         return unit->isValid();
     } ) );
 
-    const int32_t unitsToDraw = std::min( _area.width / turnOrderMonsterIconSize, validUnitCount );
+    if ( validUnitCount == 0 ) {
+        // Nothing to draw.
+        return;
+    }
+
+    auto & display = fheroes2::Display::instance();
+    _isInsideBattleField = display.height() < dialogRoi.height + turnOrderMonsterIconSize;
+
+    const int32_t unitsToDraw = std::min( _battleRoi.width / turnOrderMonsterIconSize, validUnitCount );
 
     if ( _rects.size() != static_cast<size_t>( unitsToDraw ) ) {
         // Update units icons positions.
 
-        width = turnOrderMonsterIconSize * unitsToDraw;
+        _renderingRoi.width = turnOrderMonsterIconSize * unitsToDraw;
+
+        int32_t offsetX = ( _battleRoi.width - _renderingRoi.width ) / 2;
+
+        if ( !_isInsideBattleField ) {
+            offsetX += ( display.width() - dialogRoi.width ) / 2;
```

Comment:
We can simplify the calculations:
```suggestion
            offsetX += dialogRoi.x;
```